### DISCUSSION TITLE
Survey for Mongock roadmap

### DISCUSSION THREAD

**Discussion by dieppa** [flamingock/mongock#51] [General]
Hi all, we are designing the roadmap for the next features and we think that it's crucial that we understand the user's needs and preferences. For this we have created a quick survey to understand what is important for you and also to allow you to give us any idea or need you have and we haven't thought about.

We would really appreciate you take 2 minutes to fill it.
Thanks in Advance.

[Mongock roadmap survey](https://forms.gle/uk9jmtKvxJzi5Bda7)

**Comment by dieppa** [comment]
@radhe-soni @smnnn @Endron @nicholassmith @santam85 @Zomzog  @rbraeunlich @abobwhite @dhivesh @pjonastik @vasanthraj @mbuccat @minogin @meamuri  @krasova @grabarczyk-t @glowjan @savvadem @wirefreethought @lamari @PatrykRudnicki @AleksejDomash @sriio @RameshPaul @deividhf @kascoder @nameetpai @EagleFlyInSky @xaviermichel @leopold537 @JPedroBorges @yohanliyanage @eceolin @maikdiepenbroek @j0n3 @nick318 @JosephP91 @aguiarluiz @crisaltmann @guilherme-pohlmann @ElazarNeeman @sam-element @nhomble @gchudnov @mhdzaid @mhdzaid @martin-g @juwit @CodeTripper @walczykm @vladmm

**Comment by abobwhite** [comment]
@dieppa Looks like you need to allow access to survey.

**Comment by dieppa** [comment]
@abobwhite fixed. Thank you very much

**Comment by dieppa** [comment]
@tschlegel @lukaseckert @madhbhavikar-ST @aoudiamoncef @djmpink @oubeichen @jamesjieye @olvjma @gkrupa @projected1 @phauer @Zhambul @lbcoutinho @arouel @shiyindaxiaojie @ergunp @tfij @drazzib @dogukancagatay @novboy @oackey @albertotn @tsmgeek @murdos @MattJBrowning @mxab @hananbing @hguerra @jawobar @OliRoi @alebabai @ishukhatri @hedudelgado @ashok-kallara @alansastre @unikzforce @abobwhite @engan81 @sgrillon14 @markbaird @Jim-Lin @bilelbm @kroonprins @liias @Wollender @jjangsam @Asapin @tnsnbbs @bubak4 @tnus

**Comment by dieppa** [comment]
@arnauvp @ghiron @Mrchenli @wpcfan @Danjenator @kkocel @chegola @ccamel @Hronom @cloudyrock-tech @derekbassett @zeapo @Guchman @priyank-sriv @andidev @peanutpi

**Comment by unikzforce** [comment]
It seems that you people have decided to create a very serious migration tool,
Hope your tools will be a defacto standard one day.

**Comment by dieppa** [comment]
Hi @unikzforce , yeah, we have decided to take Mongock to the next level, with the goal in mind of being the defacto open source option and providing licence for bigger companies. 

Thanks a lot for your help!

**Comment by unikzforce** [comment]
Hi there:) @dieppa  So then, you want to make it kinda polyglot? other langs aside from java? 

being a defacto standard needs a kinda separate server with some APIs for some other languages i think ( facilitated by GRPC or Openapi maybe)

**Comment by dieppa** [comment]
Ih @unikzforce ,

I actually meant to be the defacto standard for Java. However, in mid/long term, we have in mind to provide a command cli too. Go to other languages beyond Java is something that we don't have in mind right now, but it's of course a very good option in long term, but obviously, we need first to get stablished in Java and Mongo, there is a lot of work to do yet to get where we want.

However, before jumping to another programming language, we think it's more interesting for our users to provide support to other NoSql databases.

But, as always your opinion on this is much appreciated :)

**Comment by unikzforce** [comment]
@dieppa Seems very Wise, I hope I could contribute one day then

**Comment by dieppa** [comment]
And you will be more than welcome.
Thanks

**Comment by Asapin** [comment]
Thought I'll share a little bit of how we use Mongock in our company, maybe it'll be helpful for designing a roadmap.

In our case we run migrations on collections containing hundreds of thousands documents, and usually, after updating a document, we need to perform a few additional steps, like remove old version from the cache server, and send notifications to our users, that their data was updated. And because of that, our migrations tend to run for a few hours (our longest migration took about 6 hours to run).

So for us personally, having support for transactions, validations and UNDO operations doesn't really matter (we test each migration several times on different servers, before deploying it to our prod server, and we run them on every build in our CI). Instead using them would probably be harmful for us, as it increases the time needed to run the migration, and we can't really "rollback" notifications, that we already sent to users.

Also, we have many instances of each of our apps, so migrations run on whatever instance redeployed first. And in that case, having some kind of a centralized server with limited resources, that applies migrations for us and gathers some statistics (like what migrations are applied, what migration is currently being applied, maybe track how much time it took), would be very useful, as it wouldn't affect other apps. Because right now we have to manually monitor logs of all instances to pinpoint which instance is running migrations, when it started, how many documents are already processed and so on.

Some kind of UI to see what exceptions happened during migration, on what line along with a snippet of a source code with that line included, would be also really nice. Maybe even run it on that same centralized migration server?

**Comment by zeapo** [comment]
Maybe you cant take inspiration from https://flywaydb.org/

**Comment by dieppa** [comment]
Hi @zeapo , thanks a lot . We know flyaway and liquibase, we just wanted to know our user's vision on this, so we can prioritise accordingly.
 
Thanks

**Comment by dieppa** [comment]
Thanks @Asapin , we'll take that into account

**Comment by JosephP91** [comment]
I have completed the survey. I think this is the only migration tool worth to use, so keep on commit&pushing guys! 👍

**Comment by dieppa** [comment]
@JosephP91 Really thanks for your comment. Don't worry, we will :)

**Comment by dieppa** [comment]
Hello @Asapin , I understand that you run migration on high volumen of data, so you don't see any benefit on having transactions, validations, rollbacks, undo, etc.

On the other hand, you see to really benefit from migration insight and data. Not only in retrospective, but also in real time. This is one of our priorities in the roadmap, but I'd like to discuss further with you, so we can understand what data you really need.

Also you mention that you run your migration in a clustered environment, with many instances of the same app. While a centralised server would help in this area, you may loose the benefit of running the migration alongside your code. But you don't need that, you have the lock mechanism synchronising the distributed access.

I think we should bring this conversation offline. We are interested in understanding in more detail your use case. Please, feel free to send us an email to dev@cloudyrock.io so we can discuss it further

**Comment by dieppa** [comment]
@alexciornii

**Comment by gabriel-eguia** [comment]
Micronaut integration would be a really nice feature

**Comment by dieppa** [comment]
@gabriel-eguia , good point! Thanks

**Comment by MiloszKorman** [comment]
Are you considering creating a sort of a wrapper for dsl-like definition of changes in kotlin? If it doesn't interfere with your vision of mongock, I'd be interested to contribute in creation of such functionality 🙂

**Comment by dieppa** [comment]
@MiloszKorman, that sounds interesting. Unfortunately we don't have currently the resources to face that as we have already quite a lot in our roadmap.

However, I would be very happy if you want to contribute in the creation of such a feature. 
If so, please an email at antonio@cloudyrock.io and we discuss it.

**Comment by santam85** [comment]
How about a starter (as in spring-boot-starter) to manage dependencies and configuration?

**Comment by andlinger** [comment]
Quarkus integration would be really nice!

**Comment by Tcharl** [comment]
Spring Reactive support, we are struggling pretty much with [mongock at the moment](https://github.com/jhipster/generator-jhipster/pull/16787#issuecomment-1002661225)
